File: src/lib/simulation.ts
import { create } from 'zustand';
import { Job, Machine } from './types';

interface SimulationState {
  // Parameters
  jobArrivalRate: number;
  setJobArrivalRate: (rate: number) => void;
  avgJobTime: number;
  setAvgJobTime: (time: number) => void;
  numMachines: number;
  setNumMachines: (num: number) => void;
  timeScale: number;
  setTimeScale: (scale: number) => void;

  // State
  jobs: Job[];
  machines: Machine[];
  time: number;
  isRunning: boolean;
  completedJobs: Job[];
  totalQueueTime: number;
  totalServiceTime: number;
  maxQueueDisplaySize: number;
  queueSizeHistory: number[];
  utilizationHistory: number[];
  totalJobsCompleted: number;

  // Derived State (Getters)
  runningJobs: number;
  avgServiceTime: () => number;
  avgQueueTime: () => number;
  estimatedQueueEmptyTime: number;

  // Actions
  start: () => void;
  stop: () => void;
  reset: () => void;
  tick: () => void;
  updateMaxQueueDisplaySize: () => void;
}

let nextJobId = 10000;
const INITIAL_MAX_QUEUE_DISPLAY_SIZE = 50;
const QUEUE_HISTORY_LENGTH = 10;

// --- Helper Functions for Simulation Logic ---

/**
 * Processes completed jobs on machines, updating their state and collecting metrics.
 * This is crucial for ensuring that machines become available for new work and that
 * simulation statistics accurately reflect the work that has been done.
 */
const updateFinishedJobs = (machines: Machine[], time: number) => {
  let totalServiceTimeUpdate = 0;
  let totalQueueTimeUpdate = 0;
  const completedJobs: Job[] = [];

  const updatedMachines = machines.map(machine => {
    if (machine.job && machine.finishTime && time >= machine.finishTime) {
      const finishedJob = machine.job;
      const serviceTime = finishedJob.processingTime;
      const queueTime = (finishedJob.startedProcessingTime !== null) ? (finishedJob.startedProcessingTime - finishedJob.startTime) : 0;

      completedJobs.push(finishedJob);
      totalServiceTimeUpdate += serviceTime;
      totalQueueTimeUpdate += queueTime;

      return { ...machine, job: null, finishTime: null }; // Machine is now idle
    }
    return machine;
  });

  return { updatedMachines, completedJobs, totalServiceTimeUpdate, totalQueueTimeUpdate };
};

/**
 * Assigns jobs from the queue to any available and active machines.
 * This function models the core behavior of a queuing system, where work is
 * distributed to available resources.
 */
const assignJobsToMachines = (machines: Machine[], jobs: Job[], time: number) => {
  const availableMachines = machines.filter(m => m.status === 'active' && !m.job);
  const jobsToAssign = [...jobs];
  const updatedMachines = [...machines];

  for (const machine of availableMachines) {
    if (jobsToAssign.length > 0) {
      const job = jobsToAssign.shift()!;
      const jobWithStartTime = { ...job, startedProcessingTime: time };
      const machineIndex = updatedMachines.findIndex(m => m.id === machine.id);
      if (machineIndex !== -1) {
        updatedMachines[machineIndex] = {
          ...machine,
          job: jobWithStartTime,
          finishTime: time + Math.ceil(job.processingTime),
        };
      }
    }
  }

  return { updatedMachines, remainingJobs: jobsToAssign };
};

/**
 * Potentially adds a new job to the queue based on the arrival rate.
 * This simulates the inflow of work into the system, which is essential for
 * modeling real-world load.
 */
const addNewJobs = (jobs: Job[], time: number, jobArrivalRate: number, avgJobTime: number) => {
  const newJobs = [...jobs];
  if (Math.random() < jobArrivalRate / 60) {
    newJobs.push({
      id: nextJobId++,
      startTime: time,
      processingTime: Math.random() * avgJobTime * 2,
      startedProcessingTime: null,
    });
  }
  return newJobs;
};

/**
 * Removes machines that were marked for removal and are now idle.
 * This is important for dynamically scaling down the number of resources
 * without interrupting ongoing work.
 */
const removeIdleMachines = (machines: Machine[]) => {
  return machines.filter(machine => !(machine.status === 'removing' && machine.job === null));
};


export const useSimulationStore = create<SimulationState>((set, get) => ({
  // Parameters
  jobArrivalRate: 30, // Increased for better job generation
  setJobArrivalRate: (rate) => set({ jobArrivalRate: rate }),
  avgJobTime: 10,
  setAvgJobTime: (time) => set({ avgJobTime: time }),
  numMachines: 5,
  setNumMachines: (num) => {
    const currentMachines = get().machines;
    const currentNumMachines = currentMachines.length;

    if (num > currentNumMachines) {
      // Add machines
      const newMachines = [...currentMachines];
      for (let i = currentNumMachines; i < num; i++) {
        newMachines.push({ id: i, job: null, finishTime: null, status: 'active' });
      }
      set({ machines: newMachines, numMachines: num });
    } else if (num < currentNumMachines) {
      // Mark machines for removal
      const newMachines = currentMachines.map((machine, i) => {
        if (i >= num) {
          return { ...machine, status: 'removing' };
        }
        return machine;
      });
      set({ machines: newMachines, numMachines: num });
    }
  },
  timeScale: 1,
  setTimeScale: (scale) => set({ timeScale: scale }),

  // State
  jobs: [],
  machines: [],
  time: 0,
  isRunning: false,
  completedJobs: [],
  totalJobsCompleted: 0,
  totalServiceTime: 0,
  totalQueueTime: 0,
  maxQueueDisplaySize: INITIAL_MAX_QUEUE_DISPLAY_SIZE,
  queueSizeHistory: [],
  utilizationHistory: [],

  // Derived State (Getters)
  get runningJobs() {
    return get().machines.filter(machine => machine.job !== null).length;
  },
  avgServiceTime: () => {
    const { totalServiceTime, totalJobsCompleted } = get();
    return totalJobsCompleted > 0 ? totalServiceTime / totalJobsCompleted : 0;
  },
  avgQueueTime: () => {
    const { totalQueueTime, totalJobsCompleted } = get();
    return totalJobsCompleted > 0 ? totalQueueTime / totalJobsCompleted : 0;
  },

  // Actions
  start: () => set({ isRunning: true }),
  stop: () => set({ isRunning: false }),
  reset: () => {
    const machines: Machine[] = [];
    for (let i = 0; i < get().numMachines; i++) {
      machines.push({ id: i, job: null, finishTime: null, status: 'active' });
    }
    set({
      jobs: [],
      machines,
      time: 0,
      isRunning: false,
      completedJobs: [],
      totalJobsCompleted: 0,
      totalServiceTime: 0,
      totalQueueTime: 0,
      maxQueueDisplaySize: INITIAL_MAX_QUEUE_DISPLAY_SIZE,
      queueSizeHistory: [],
      utilizationHistory: [],
    });
  },
  tick: () => {
    if (!get().isRunning) return;

    const state = get();
    const newNow = state.time + 1;

    // --- Simulation Steps ---

    // 1. Update finished jobs
    const { updatedMachines: machinesAfterCompletion, completedJobs, totalServiceTimeUpdate, totalQueueTimeUpdate } = updateFinishedJobs(state.machines, newNow);

    // 2. Remove idle machines marked for removal
    const machinesAfterRemoval = removeIdleMachines(machinesAfterCompletion);

    // 3. Assign jobs to available machines
    const { updatedMachines: machinesAfterAssignment, remainingJobs } = assignJobsToMachines(machinesAfterRemoval, state.jobs, newNow);

    // 4. Add new jobs to the queue
    const newJobs = addNewJobs(remainingJobs, newNow, state.jobArrivalRate, state.avgJobTime);

    const utilization = state.numMachines > 0 ? (machinesAfterAssignment.filter(m => m.job).length / state.numMachines) * 100 : 0;

    // --- Update State ---
    set(prevState => ({
      time: newNow,
      machines: machinesAfterAssignment,
      jobs: newJobs,
      completedJobs: [...prevState.completedJobs, ...completedJobs],
      totalJobsCompleted: prevState.totalJobsCompleted + completedJobs.length,
      totalServiceTime: prevState.totalServiceTime + totalServiceTimeUpdate,
      totalQueueTime: prevState.totalQueueTime + totalQueueTimeUpdate,
      queueSizeHistory: [...prevState.queueSizeHistory.slice(-(QUEUE_HISTORY_LENGTH - 1)), newJobs.length],
      utilizationHistory: [...prevState.utilizationHistory, utilization],
    }));

    get().updateMaxQueueDisplaySize();
  },
  updateMaxQueueDisplaySize: () => {
    const state = get();
    const currentQueueSize = state.jobs.length;
    let newMaxQueueDisplaySize = state.maxQueueDisplaySize;

    // Resizing Up Logic
    const historyThreshold = Math.floor(QUEUE_HISTORY_LENGTH * 0.8); // e.g., 80% of history
    const consistentlyAboveThreshold = state.queueSizeHistory.filter(size => size > newMaxQueueDisplaySize).length >= historyThreshold;

    if (currentQueueSize > newMaxQueueDisplaySize && consistentlyAboveThreshold) {
      newMaxQueueDisplaySize *= 2;
    }

    // Resizing Down Logic
    const consistentlyBelowThreshold = state.queueSizeHistory.filter(size => size < newMaxQueueDisplaySize / 4).length >= historyThreshold; // e.g., below 25% of current max

    if (currentQueueSize < newMaxQueueDisplaySize / 2 && consistentlyBelowThreshold && newMaxQueueDisplaySize > INITIAL_MAX_QUEUE_DISPLAY_SIZE) {
      newMaxQueueDisplaySize /= 2;
    }

    if (newMaxQueueDisplaySize !== state.maxQueueDisplaySize) {
      set({ maxQueueDisplaySize: newMaxQueueDisplaySize });
    }
  },
}));

File: python/simulation.py
import random
import math
from typing import List
from .models import Job, Machine, SimulationConfig

class SimulationEngine:
    def __init__(self, config: SimulationConfig):
        self.config = config
        self.time = 0
        self.is_running = False
        self.jobs_in_queue: List[Job] = []
        self.machines: List[Machine] = [Machine(id=i) for i in range(config.num_machines)]
        self.completed_jobs_count = 0
        self.next_job_id = 10000

        # History for graphs
        self.utilization_history: List[float] = []
        self.queue_size_history: List[int] = []
        self.needed_machines_history: List[int] = []

        # Metrics
        self.total_queue_time = 0
        self.total_service_time = 0

    def reset(self):
        self.time = 0
        self.is_running = False
        self.jobs_in_queue = []
        self.machines = [Machine(id=i) for i in range(self.config.num_machines)]
        self.completed_jobs_count = 0
        self.next_job_id = 10000
        self.utilization_history = []
        self.queue_size_history = []
        self.needed_machines_history = []
        self.total_queue_time = 0
        self.total_service_time = 0

    def tick(self):
        if not self.is_running:
            return

        self.time += 1

        # 1. Process finished jobs
        for machine in self.machines:
            if machine.job and machine.finish_time and machine.finish_time <= self.time:
                # Job finished
                self.completed_jobs_count += 1
                self.total_queue_time += (machine.job.started_processing_time - machine.job.start_time)
                self.total_service_time += machine.job.processing_time
                machine.job = None
                machine.finish_time = None

        # 2. Decommission machines (not implemented in this version as per spec)

        # 3. Assign new jobs
        for machine in self.machines:
            if not machine.job and self.jobs_in_queue:
                job_to_assign = self.jobs_in_queue.pop(0)
                job_to_assign.started_processing_time = self.time
                machine.job = job_to_assign
                machine.finish_time = self.time + job_to_assign.processing_time

        # 4. Generate new jobs
        if random.random() < self.config.job_arrival_rate / 60:
            new_job = Job(
                id=self.next_job_id,
                start_time=self.time,
                processing_time=int(random.uniform(self.config.avg_job_time / 2, self.config.avg_job_time * 1.5))
            )
            self.jobs_in_queue.append(new_job)
            self.next_job_id += 1

        # 5. Update historical data
        running_jobs = sum(1 for m in self.machines if m.job)
        utilization = (running_jobs / len(self.machines)) * 100 if self.machines else 0
        self.utilization_history.append(utilization)
        self.queue_size_history.append(len(self.jobs_in_queue))

        total_remaining_processing_time = sum(j.processing_time for j in self.jobs_in_queue)
        for m in self.machines:
            if m.job and m.finish_time:
                total_remaining_processing_time += (m.finish_time - self.time)
        
        estimated_time = total_remaining_processing_time / len(self.machines) if self.machines else 0
        needed_machines = math.ceil(total_remaining_processing_time / 60) - len(self.machines) if estimated_time > 60 else 0
        self.needed_machines_history.append(max(0, needed_machines))

    def update_config(self, new_config: SimulationConfig):
        # Adjust number of machines
        if new_config.num_machines > len(self.machines):
            for i in range(len(self.machines), new_config.num_machines):
                self.machines.append(Machine(id=i))
        elif new_config.num_machines < len(self.machines):
            # Mark machines for removal (simplified: just remove idle ones)
            num_to_remove = len(self.machines) - new_config.num_machines
            idle_machines = [m for m in self.machines if not m.job]
            for i in range(min(num_to_remove, len(idle_machines))):
                self.machines.remove(idle_machines[i])

        self.config = new_config
